src/term/mod.rs

```diff
@@ -87,15 +127,32 @@ impl<'a> Drop for RenderableCellsIter<'a> {
     /// Resets temporary render state on the grid
     fn drop(&mut self) {
         if self.cursor_is_visible() {
-            self.grid[self.cursor].swap_fg_and_bg();
+            let cursor_cell = match self.cursor_cell {
+                CursorCell::Extra(ref mut cell) => cell,
+                CursorCell::Replacement(ref mut cell) => cell,
+            };
+            match self.cursor_type.0 {
+                ansi::CursorType::BlinkBlock => {
+                    cursor_cell.inner.swap_fg_and_bg();
+                }
+                ansi::CursorType::SteadyBlock => {
+                    cursor_cell.inner.swap_fg_and_bg();
+                }
+                ansi::CursorType::BlinkUnderline => {}
+                ansi::CursorType::SteadyUnderline => {}
+                ansi::CursorType::BlinkBeam => {}
+                ansi::CursorType::SteadyBeam => {
+                    cursor_cell.inner = self.grid[self.cursor];
+                }
+            }
         }
     }
 }
 
 pub struct IndexedCell {
     pub line: Line,
     pub column: Column,
-    pub inner: Cell
+    pub inner: Cell,
```

Comment:
There's a bunch of stylistic changes in here that make it difficult to review the parts that are actually relevant to your feature.